gluten-core/src/main/scala/org/apache/gluten/utils/PullOutProjectHelper.scala

```diff
@@ -143,8 +144,39 @@ trait PullOutProjectHelper {
     ae.copy(aggregateFunction = newAggFunc, filter = newFilter)
   }
 
+  private def needPreComputeRangeFrameBoundary(bound: Expression): Boolean = {
+    bound match {
+      case _: PreComputeRangeFrameBound => false
+      case _ if !bound.foldable => false
+      case _ => true
+    }
+  }
+
+  private def preComputeRangeFrameBoundary(
+      bound: Expression,
+      orderSpec: SortOrder,
+      expressionMap: mutable.HashMap[Expression, NamedExpression]): Expression = {
+    bound match {
+      case _: PreComputeRangeFrameBound => bound
+      case _ if !bound.foldable => bound
+      case _ if bound.foldable =>
+        val a = expressionMap
+          .getOrElseUpdate(
+            bound.canonicalized,
+            Alias(Add(orderSpec.child, bound), generatePreAliasName)())
```

Comment:
Do we really support reuse the reference for the same bound literal ? What happens if the two window expressions have the same frame but requires different ordering column.
```
  sum(x) OVER (
    PARTITION BY c1
    ORDER BY c2 RANGE BETWEEN 1 preceding AND CURRENT ROW
  ), 
  sum(x) OVER (
    PARTITION BY c1
    ORDER BY c3 RANGE BETWEEN 1 preceding AND CURRENT ROW
  )
```

Reply:
> Do we really support reuse the reference for the same bound literal ? What happens if the two window expressions have the same frame but requires different ordering column.
> 
> ```
>   sum(x) OVER (
>     PARTITION BY c1
>     ORDER BY c2 RANGE BETWEEN 1 preceding AND CURRENT ROW
>   ), 
>   sum(x) OVER (
>     PARTITION BY c1
>     ORDER BY c3 RANGE BETWEEN 1 preceding AND CURRENT ROW
>   )
> ```

let me check

Reply:
The frame bouding is a reference, which will be pre-project before window operator. And the sort/partition are calculated in window operator. So it's ok to reuse it.
I add a UT for this.

Reply:
The new added UT succeed in spark3.2, but failed in spark3.4/3.5. I'll fix it

Reply:
Sorry for late response, all UT passed now, the failed one seems unrelated. Please help review again @PHILO-HE @ulysses-you 

The reason why it was failed is because  sort column `l_discount` is a decimal, and when we do `Add(sort_col, offset)`, the precision of decimal type will change, so that the data type check failed in Spark (https://github.com/apache/spark/blob/82a84ede6a47232fe3af86672ceea97f703b3e8a/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/windowExpressions.scala#L83)

In fact, we only support `ByteType | ShortType | IntegerType | LongType | DateType` by design, which is checked in https://github.com/apache/incubator-gluten/pull/5626/files#diff-68925b6a5ee67a942a976971ae6727a45a0b166d200d8d983454169e7c3bd1f6R316, but even when WindowExec fallback by this check, the `PullOutPreProjection` rule will still be executed.